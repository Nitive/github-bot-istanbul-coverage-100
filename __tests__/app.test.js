const { run } = require('../app')
const { baseOctokit } = require('./octokit')

const octokit = baseOctokit
  .extend('POST /repos/:owner/:repo/check-runs', () => Promise.resolve())
  .extend('GET /repos/:owner/:repo/issues/:issueNumber/comments', () => Promise.resolve({ data: [] }))
  .extend('POST /repos/:owner/:repo/issues/:issueNumber/comments', () => Promise.resolve())

const env = {
  TRAVIS_PULL_REQUEST: '0',
  TRAVIS_PULL_REQUEST_SHA: 'sha',
}

const badCoverageReport = {
  total: {
    lines: { pct: 92.61 },
    statements: { pct: 91.99 },
    functions: { pct: 76.83 },
    branches: { pct: 93.75 },
  },
}

const goodCoverageReport = {
  total: {
    lines: { pct: 100 },
    statements: { pct: 100 },
    functions: { pct: 100 },
    branches: { pct: 100 },
  },
}

describe('App', () => {
  it('should set failed check when coverage is below 100%', async () => {
    expect.assertions(3)

    await run({
      octokit: octokit.extend('POST /repos/:owner/:repo/check-runs', (result) => {
        expect(result).toMatchObject({
          name: 'Coverage',
          conclusion: 'failure',
          status: 'completed',
        })
        expect(result.output).toMatchObject({
          title: 'ðŸ’” below 100%',
        })
        expect(result.output.summary).toMatchSnapshot()

        return Promise.resolve()
      }),
      env,
      report: badCoverageReport,
    })
  })

  it('should set success check when coverage is 100%', async () => {
    expect.assertions(3)

    await run({
      octokit: octokit.extend('POST /repos/:owner/:repo/check-runs', (result) => {
        expect(result).toMatchObject({
          name: 'Coverage',
          conclusion: 'success',
          status: 'completed',
        })
        expect(result.output).toMatchObject({
          title: 'ðŸ’š Everything is covered',
        })
        expect(result.output.summary).toMatchSnapshot()

        return Promise.resolve()
      }),
      env,
      report: goodCoverageReport,
    })
  })

  it('should not leave comment when coverage is 100%', async () => {
    const createComment = jest.fn()

    await run({
      octokit: octokit.extend('POST /repos/:owner/:repo/issues/:issueNumber/comments', () => {
        createComment()
        return Promise.resolve()
      }),
      env,
      report: goodCoverageReport,
    })

    expect(createComment).not.toHaveBeenCalled()
  })
})
