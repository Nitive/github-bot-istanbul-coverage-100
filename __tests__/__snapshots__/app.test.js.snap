// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should leave comment when coverage is below 100% 1`] = `
Object {
  "body": "<!-- commentType: \\"coverage-report\\" -->
### ğŸ’” Coverage report

Type | Coverage
:-|-:
Statements | ğŸ’” \`91.99%\`
Branches | ğŸ’” \`93.75%\`
Functions | ğŸ’” \`76.83%\`
Lines | ğŸ’” \`92.61%\`",
  "headers": Object {
    "accept": "application/vnd.github.machine-man-preview+json",
    "authorization": "token installation-token",
  },
  "issueNumber": "1",
  "owner": "Nitive",
  "repo": "github-bot-istanbul-coverage-100",
}
`;

exports[`App should set failed check when coverage is below 100% 1`] = `
"<!-- commentType: \\"coverage-report\\" -->
### ğŸ’” Coverage report

Type | Coverage
:-|-:
Statements | ğŸ’” \`91.99%\`
Branches | ğŸ’” \`93.75%\`
Functions | ğŸ’” \`76.83%\`
Lines | ğŸ’” \`92.61%\`"
`;

exports[`App should set success check when coverage is 100% 1`] = `
"<!-- commentType: \\"coverage-report\\" -->
### ğŸ’š Coverage report

Type | Coverage
:-|-:
Statements | ğŸ’š \`100.0%\`
Branches | ğŸ’š \`100.0%\`
Functions | ğŸ’š \`100.0%\`
Lines | ğŸ’š \`100.0%\`"
`;
